### DISCUSSION TITLE
Help for Translation Support

### DISCUSSION THREAD

**Discussion by SardanDev** [Chatterino/chatterino2#4508] [Ideas]
Hello,

as very kindly suggested by @pajlada. I would like to contribute to this project and particularly to the translation support.

**Comment by pajlada** [comment]
Open questions:

- How do we change our strings to be translatable with Qt? OBS might be a good source for inspiration
- How do we handle people adding translations? OBS and DankChat might be good sources for inspiration

**Reply by SardanDev** [reply]
So, unfortunately I'm not really a software developer but more of add-ons or bots in javascript or websites. I was proposing my help more for the translation itself but really for the integration of this one of it but I can find out with the informations you gave.

**Reply by Nerixyz** [reply]
> How do we change our strings to be translatable with Qt?

I've looked at some other Qt projects (mainly ones from the [`qt` GitHub topic](https://github.com/topics/qt)):

## Custom Solutions

### [OBS](https://github.com/obsproject/obs-studio)

OBS uses the translation capabilities of their underlying library `libobs` ([docs](https://docs.obsproject.com/reference-libobs-util-text-lookup)). Translations are stored in `.ini` files ([example](https://github.com/obsproject/obs-studio/blob/13dc9a2b0db0f0a2f74cc6aebb4647544ebe3f5c/UI/data/locale/en-GB.ini)). Formatting is done with `QString::arg` (a string is specified like `%1 apples`). A string is fetched using the `QTStr` macro (expands to `QString`) or using `Str` function (regular `char*`) ([source](https://github.com/obsproject/obs-studio/blob/13dc9a2b0db0f0a2f74cc6aebb4647544ebe3f5c/UI/obs-app.hpp#L249-L253)).

### [Telegram](https://github.com/telegramdesktop/tdesktop)

Telegram uses their own translation framework. They're using their own [`codegen`](https://github.com/desktop-app/codegen/) to generate code for translatable strings. The format tightly integrates with their other libraries, such as [`lib_rpl`](https://github.com/desktop-app/lib_rpl).

## [GNU `gettext`](https://www.gnu.org/software/gettext/manual/html_node/index.html)

* [VLC (Qt)](https://code.videolan.org/videolan/vlc/-/tree/master/modules/gui) - [`/po` directory](https://code.videolan.org/videolan/vlc/-/blob/d02a131e03960d364ac73ef1cba4ca03e876b626/po) (translations are powered by the underlying `libvlc`)
* [spyder](https://github.com/spyder-ide/spyder) - [`/locale` directory](https://github.com/spyder-ide/spyder/tree/37301078fd5a3149ace619b03dd15aae0459f944/spyder/locale)

## [Qt's Internationalization](https://doc.qt.io/qt-6/internationalization.html)

* [FreeCAD](https://github.com/FreeCAD/FreeCAD) - [`/Language` directory](https://github.com/FreeCAD/FreeCAD/tree/fbee0c92b602ab4efb700ea2e1d34d00f3225e34/src/Gui/Language)
* [MEGAsync](https://github.com/meganz/MEGAsync) - [`/translations` directory](https://github.com/meganz/MEGAsync/tree/bd02362d0622510969197f3fcd1005b869506dd8/src/MEGASync/gui/translations)
* [flameshot](https://github.com/flameshot-org/flameshot) - [`/translations` directory](https://github.com/flameshot-org/flameshot/tree/master/data/translations)
* [tiled](https://github.com/mapeditor/tiled) - [`/translations` directory](https://github.com/mapeditor/tiled/tree/358b6a2d21cdcf12b006706d43ee17d64fa1bb97/translations)
* [MuseScore](https://github.com/musescore/MuseScore) - [`/locale` directory](https://github.com/musescore/MuseScore/tree/062dae1b0b034ce1304e43160577a2fe73beaf9b/share/locale)
* [QtScrcpy](https://github.com/barry-ran/QtScrcpy) - [`/i18n` directory](https://github.com/barry-ran/QtScrcpy/tree/cec31a6f675e9a00bc4a56db96b7cfbc856d797b/QtScrcpy/res/i18n)
* [liteide](https://github.com/visualfc/liteide) - [directory](https://github.com/visualfc/liteide/tree/abe3def1be143a434007806b5371a2451795a34f/liteidex)
* [olive](https://github.com/olive-editor/olive) - [`/ts` directory](https://github.com/olive-editor/olive/tree/5626e6aeb1295719bf151cd94129367549f32cd1/app/ts)
* [NotepadNext](https://github.com/dail8859/NotepadNext) - [`/i18n` directory](https://github.com/dail8859/NotepadNext/tree/5b1583112fa18780d2f67b821f61d2ab9533779e/i18n)
* [moonlight-qt](https://github.com/moonlight-stream/moonlight-qt) - [`/languages` directory](https://github.com/moonlight-stream/moonlight-qt/tree/68daa90b7eeda2bbc3eeb55d53ba0af831e4c416/app/languages)
* [qTox](https://github.com/qTox/qTox) (archived) - [`/translations` directory](https://github.com/qTox/qTox/tree/8632be21bebfa9ece815c3af4e298b775f50c85e/translations)
